core-to-plc/src/Language/Plutus/CoreToPLC/Plugin.hs

```diff
@@ -0,0 +1,177 @@
+      GHC.Cast e c -> flip GHC.Cast c <$> conv e
+      GHC.Tick t e -> GHC.Tick t <$> conv e
+      e@(GHC.Coercion _) -> return e
+      e@(GHC.Lit _) -> return e
+      e@(GHC.Var _) -> return e
+      e@(GHC.Type _) -> return e
+
+data CoreToPlcFailure = CoreToPlcFailure
+
+instance Show CoreToPlcFailure where
+    show _ = "Core to PLC plugin failure"
+
+instance Exception CoreToPlcFailure
+
+-- | Actually invokes the Core to PLC compiler to convert an expression into a PLC literal.
+convertExpr :: GHC.CoreExpr -> GHC.Type -> GHC.CoreM GHC.CoreExpr
+convertExpr origE tpe = do
+    flags <- GHC.getDynFlags
+    primTerms <- makePrimitiveMap primitiveTermAssociations
+    primTys <- makePrimitiveMap primitiveTypeAssociations
+    let result =
+          do
+              converted <- convExpr origE
+              -- temporarily don't do typechecking due to lack of support for redexes
+              --annotated <- convertErrors PCError $ PC.annotateTerm converted
+              --inferredType <- convertErrors PCError $ PC.typecheckTerm 1000 annotated
+              pure (converted, undefined)
+    case runExcept $ runReaderT (runQuoteT result) (flags, primTerms, primTys, initialScopeStack) of
+        Left s -> do
+            GHC.fatalErrorMsg $
+                "Failed to convert GHC core expression:" GHC.$+$
+                GHC.ppr origE GHC.$+$
+                "Error is:" GHC.$+$
+                (GHC.text $ T.unpack $ errorText s)
+            -- this will actually terminate compilation
+            liftIO $ throwIO CoreToPlcFailure
+        Right (term, _) -> do
+            let termRep = T.unpack $ PC.debugText term
+            --let typeRep = T.unpack $ PC.debugText inferredType
+            -- Note: tests run with --verbose, so these will appear
+            GHC.debugTraceMsg $
+                "Successfully converted GHC core expression:" GHC.$+$
+                GHC.ppr origE GHC.$+$
+                "Resulting PLC term is:" GHC.$+$
+                GHC.text termRep --GHC.$+$ "With type:" GHC.$+$ GHC.text typeRep
+            let program = PC.Program () (PC.defaultVersion ()) term
+            let serialized = PC.writeProgram program
+            -- The GHC api only exposes a way to make literals for Words, not Word8s, so we need to convert them
+            let (word8s :: [Word]) = fromIntegral <$> BSL.unpack serialized
+            -- The flags here are so GHC can check whether the word is in range for the current platform.
+            -- This will never actually be a problem for us, since they're really Word8s, but GHC
+            -- doesn't know that.
+            let (wordExprs :: [GHC.CoreExpr]) = fmap (GHC.mkWordExprWord flags) word8s
+            let listExpr = GHC.mkListExpr GHC.wordTy wordExprs
+            -- Our result type is representationally equivalent to [Word],
+            -- so we can insert a simple coercion here
+            pure $ GHC.Cast listExpr $ GHC.mkRepReflCo tpe
```

Comment:
Minor nitpick: is there any logic to where you use `pure` and where `return`.

Reply:
I intend to use `pure` everywhere, but indeed I've somehow got a few `return`s.